### PR TITLE
Require commit SHA for /test and /test-extended on external PRs

### LINKED ISSUE #1538: Tidy-up/polish PR comment bot implementation
Following on from the implementation work for #478, there are some aspects of the implementation that could be tidied up and polished. This includes:

- [x] Move GitHub script code to external js file for better editing experience in VSCode (#1616 )
- [x] Add `/test-extended` command to also run the extended tests (#1576 )
- [x] Add tests for the GitHub script code (#1660)
- [x] When running `/test-destroy-env`, run the destroy twice: once generating the environment ID for the PR and once for the branch. This way, we can clean up  both environments. (note: the destroy needs to gracefully handle the environment not existing) (#1663 )
- [x] Consider splitting the cleanup step to a separate job. This would be clearer when looking at the job run history, as well as limiting the scope of the workflow requiring credentials (implemented in #1663 )
- [x] Change command test logic: Trim first line and check for leading `/`. If found, assume a command is intended and continue to match. If not match then output help message `'/foo' not recognised. Available commands are ...` (implemented in #1727)
- [x] Add a comment when a user without permissions attempts to run commands (implemented in #1727)
- [x] Show who left the comment with the command that the bot is responding to in any comments the bot leaves in response (implemented in #1727)
- [x] Have the bot add a comment in response to `/test-force-approve` commands indicating the SHA being approved and the original commenter (implemented in #1727)
- [x] Include the resource group name (or at least the pr ref SHA) in bot comments for test runs to make it easier when troubleshooting runs (implemented in #1727)
- [ ] require head SHA on external PRs for `/test` (#1731)
- [ ] Ignore `/test` etc on closed PRs? (Possibly still respond to `/test-destroy-env`?). (#1731)<br>Update: running `/test` on closed PRs fails as there's no `potentialMergeCommit` generated by GitHub. <br>Update 2: running `/test` on open PRs with merge conflicts fails as there's no `potentialMergeCommit` generated

Other aspects that could be considered for tidy-up:
- Update 3rd party actions to use specific (validated) SHAs
- addressing linter warnings for the workflow YAML files
- ~~In `deploy_tre_reusable.yml`, move non-sensitive fields (e.g. `TRE_ID` to `input` rather than `secret`) to improve log output by removing unnecessary masking (Initially attempted in #1544 but later reverted)~~

### PR DESCRIPTION
**Pull request by stuartleeks**
- [x] Test improvements (more consistency in structure, add helpers to make tests more readable)
- [x] Update `/test` and `/test-extended` to require SHA of reviewed commit for external PRs
- [x] Add `--watch` option to `run-tests.sh` scrip
- [x] Update `maintainers.md` for command changes
- [x] Make pr-bot tolerant of spaces between command and SHA
- [x] Add comment and stop for test commands on non-mergeable PRs (i.e. without potential merge commit to check out)

Closes #1538

### CODE REVIEW

File: .github/scripts/build.js [hunk 1]
```diff
@@ -57,33 +65,39 @@ async function getCommandFromComment({ core, context, github }) {
...
+    const commandText = parts[0];
+    switch (commandText) {
       case "/test":
         {
-          if (gotNonDocChanges) {
-            command = "run-tests";
-            const message = `:runner: Running tests: https://github.com/${repoFullName}/actions/runs/${runId} (with refid \`${prRefId}\`)`;
-            await addActionComment({ github }, repoOwner, repoName, prNumber, commentUsername, commentLink, message);
-          } else {
+          // Docs only changes don't run tests with secrets so can run regardless of whether
```
Comment (martinpeck) [inline .github/scripts/build.js:73]: Is this sentence missing and ending?

File: .github/scripts/build.js [hunk 2]
```diff
@@ -57,33 +65,39 @@ async function getCommandFromComment({ core, context, github }) {
...
+          if (!gotNonDocChanges) {
```
Comment (martinpeck) [inline .github/scripts/build.js:74]: I know this isn't introduced in your PR, but I struggle a little with the name of this variable.

You're basically testing "is this a doc-only change". Could the variable be called `docOnlyChange` or `nonCodeChange`?

This is a nit. Feel free to ignore.

File: .github/scripts/build.js [hunk 3]
```diff
@@ -108,6 +122,47 @@ async function getCommandFromComment({ core, context, github }) {
...
+  const externalPr = !prAuthorHasWriteAccess;
+  if (externalPr) {
+    if (commandParts.length === 1) {
+      const message = `:warning: When using \`${command}\` on external PRs, the SHA of the checked commit must be specified`;
+      await addActionComment({ github }, pr.repoOwner, pr.repoName, pr.number, comment.username, comment.link, message);
+      return false;
+    }
+    const commentSha = commandParts[1];
+    if (commentSha.length < 7) {
+      const message = `:warning: When specifying a commit SHA it must be at least 7 characters (received \`234567\`)`;
```
Comment (martinpeck) [inline .github/scripts/build.js:149]: is `234567` supposed to be the commentSha?

**Review by martinpeck — APPROVED** [review]

File: .github/scripts/build.js (hunk 1 above)
Comment (stuartleeks) [reply .github/scripts/build.js]: Oops, good spot - fixed

File: .github/scripts/build.js (hunk 2 above)
Comment (stuartleeks) [reply .github/scripts/build.js:75]: Yeah, I agree - updated

File: .github/scripts/build.js (hunk 3 above)
Comment (stuartleeks) [reply .github/scripts/build.js]: Ooops - update to insert the correct value!

**Comment by stuartleeks** [comment]
/test-force-approve

pr-bot tests don't execute these changes + have run test suite locally since making PR feedback updates

File: .github/scripts/build.js (hunk 2 above)
Comment (stuartleeks) [reply .github/scripts/build.js:75]: oops - realised that I forgot to push this commit. Can pick this up in any future changes in the code!

**Pull request merged**